PROD-145: rework scope and scope restrictions to deny access base on rules

icehess opened the pull request:
This will update scope functionality in a way that all application can use.

When creating JWT, modules all defined scope on request (if allowed), user doc, account doc and available scopes from system will be added to JET token.

During authorization, application can call `kz_auth_scope:maybe_deny_access(AuthDoc, Options)` to check if request is authorized based on scopes defined in the token and their rules.

Scope rules (scope restrictions) are defined for each scope in a document system wide or in account db. The rules may support multiple transport. Currently we have `http`. Other possible transport like `websocket`, `sip` may be added in the future if needed.

Here is HTTP scope restrictions sample:
```json
{
   "_id":"scope:read_only",
   "scope_name": "read_only",
   "http":{
      "allow":{
            "/v2/accounts/*/devices/0fbd9844134db44cac13b365b5f384f9"
      },
      "deny":{
         "post":[
            "/v2/accounts/*/devices/*",
            "/v2/accounts/*/users/*",
            "/v2/accounts/*/callflows/*"
         ],
         "patch":[
            "*"
         ]
      }
   },
   "precedence":"allow",
   "pvt_doc_type":"scope_restriction"
}
```

icehess committed:
add and scope to auth token and validate scope restrictions

icehess committed:
auth http transport, move restritions to scope module

icehess committed:
check if scope can be request from api and cleanup scope view

icehess committed:
remove unused scope restriction view

icehess committed:
update default scopes doc

icehess committed:
small refactoring to use function

icehess committed:
prefi user type with user: when adding it to scope claim

icehess committed:
make apis

icehess commented:
Circular Required PR:
https://github.com/2600hz/kazoo-crossbar/pull/163

icehess committed:
fix view and dialyzer

lazedo commented on kazoo_auth/priv/couchdb/views/scopes.json:
```diff
@@ -13,22 +13,18 @@
-                "  emit(doc._id, doc.scopes);",
+                "  emit(doc._id, {",
+                "    'scopes': doc.scopes,",
+                "    'allow_api_request': doc.allow_api_request || true",
```
why is this changing ? `allow_api_request` seems specific to `crossbar`

lazedo commented on kazoo_auth/priv/couchdb/views/scopes.json:
```diff
@@ -13,22 +13,18 @@
-                "      });",
-                "    }",
-                "  }",
+                "  emit([doc.allow_api_request || true, doc._id], doc.scopes);",
```
same as above. `allow_api_request`  seems specific to `crossbar`

lazedo replied:
why the renaming of the view and changing what it emits ?

lazedo commented on kazoo_auth/src/kz_auth.erl:
```diff
@@ -187,6 +188,13 @@ include_identity_sign(Claims) ->
+        {'ok', NewClaim} -> NewClaim;
+        _Else -> Claims
+    end.
+
```
why is this included here ?

lazedo commented on kazoo_auth/src/kz_auth.erl:
```diff
@@ -173,6 +173,7 @@ ensure_claims(#{payload := Payload}) ->
 -spec include_claims(kz_term:proplist()) -> kz_term:proplist().
 include_claims(Claims) ->
     Routines = [fun include_identity_sign/1
+               ,fun include_scope/1
```
why is this included here ? it seems it will not honor the scopes requested to be emitted by caller

lazedo commented on kazoo_auth/src/kz_auth_scope.erl:
```diff
@@ -17,58 +17,78 @@
         ,available_scopes/0, available_scopes/1
         ,exists/1
-        ,to_list/1
-        ,to_str/1
```
why is this changing ? these are simple `transformation` functions, renaming to `join` & `split` seems more confusing on the intent.

lazedo commented on kazoo_auth/src/kz_auth.erl:
```diff
@@ -187,6 +188,13 @@ include_identity_sign(Claims) ->
+-spec include_scope(kz_term:proplist()) -> kz_term:proplist().
+include_scope(Claims) ->
+    case kz_auth_scope:create_claims(Claims) of
+        {'ok', NewClaim} -> NewClaim;
```
why are the original claims being overridden ?

lazedo commented on kazoo_auth/src/kz_auth_scope.erl:
```diff
@@ -77,39 +97,114 @@ has_scopes(Token, ScopeList, Option) ->
+            {'error', kz_binary:format("failed to open doc ~s: ~p", [AccountId, _Reason])}
+    end;
+get_account_scopes(_) ->
+    {'ok', []}.
```
why isn't this done in crossbar before calling `kz_auth:create_token/1` ?

lazedo commented on kazoo_auth/src/kz_auth_scope.erl:
```diff
@@ -17,58 +17,78 @@
 -define(SCOPE_DELIM, <<":">>).
 -define(SCOPE_SEP, <<" ">>).
--define(SCOPES_KEY, <<"scopes">>).
--define(SCOPES_CAT, <<"system_auth">>).
--define(SCOPES_LIST, <<"scopes/crossbar_listing">>).
```
why is this removed ? relocating the db is easier this way instead of going thru the code.

lazedo commented on kazoo_auth/src/kz_auth_scope.erl:
```diff
@@ -77,39 +97,114 @@ has_scopes(Token, ScopeList, Option) ->
+    [];
+allowed_request_scopes(Scopes) ->
+    Options = [{'keys', Scopes}],
+    case kz_datamgr:get_results(?KZ_AUTH_DB, <<"scopes/list_allowed_by_scope_id">>, Options) of
```
see comment about the removed `macro`

lazedo replied:
putting this here will override intended claims & scopes passed in call to kz_auth:create_token/1

lazedo commented on kazoo_auth/src/kz_auth_scope.erl:
```diff
@@ -77,39 +97,114 @@ has_scopes(Token, ScopeList, Option) ->
-    case kz_datamgr:get_results(?SCOPES_CAT, View, ViewOptions) of
+-spec fetch_scopes(kazoo_data:view_options()) -> kz_term:ne_binaries().
+fetch_scopes(ViewOptions) ->
+    case kz_datamgr:get_results(?KZ_AUTH_DB, <<"scopes/crossbar_listing">>, ViewOptions) of
```
see comment about removed `macro`

lazedo commented on kazoo_auth/src/kz_auth_scope.erl:
```diff
@@ -77,39 +97,114 @@ has_scopes(Token, ScopeList, Option) ->
               ,'include_docs'
               ],
-    Scopes = fetch_scopes(?SCOPES_LIST, Options),
+    Scopes = fetch_scopes(Options),
```
i understand this is the result of view rename and key changing, but why ?

lazedo commented on kazoo_auth/src/kz_auth_scope.erl:
```diff
@@ -215,45 +308,116 @@ resolve_scopes_fold(Scope, Available) ->
+        'false' ->
+            lager:debug("invalid scope transport ~s", [Transport]),
+            'false'
+    end.
```
why isn't this done in crossbar ?

lazedo commented on kazoo_auth/src/kz_auth_scope.erl:
```diff
@@ -215,45 +308,116 @@ resolve_scopes_fold(Scope, Available) ->
+
+-spec restriction_doc_id(kz_term:ne_binary()) -> kz_term:ne_binary().
+restriction_doc_id(<<?SCOPE_RESTRICTION_PREFIX, _/binary>> = Id) -> Id;
+restriction_doc_id(Scope) -> <<?SCOPE_RESTRICTION_PREFIX, Scope/binary>>.
```
separation of concerns here. kz_auth_scope shouldn't deal with restrictions

lazedo commented on kazoo_auth/src/kz_auth_transport_http.erl:
```diff
@@ -0,0 +1,81 @@
+    of
+        'false' -> is_match_endpoint(PathParam, Rules);
+        'true' -> 'true'
+    end.
```
why wasn't the restriction logic added to `crossbar` ?

lazedo commented on kazoo_auth/src/kz_auth_scope.erl:
```diff
@@ -17,58 +17,78 @@
+        ]).
+
+-define(SCOPE_PREFIX, "scope:").
+-define(SCOPE_RESTRICTION_PREFIX, "scope_rule:").
```
not sure this works well with couch3, i think the `:` has special meaning in couchdb3 docid for partitioned databases, check [here](https://docs.couchdb.org/en/stable/partitioned-dbs/index.html)

lazedo reviewed (commented):
a lot of things added in `core` that imo belong to crossbar or other callers of `kz_auth:create_token/1`. with this approach, if you try to create a token with userid/account_id with explicit scopes you will always get extra scopes that are not intended.

`kz_auth_scope` should handle aggregation & verification of scopes, it shouldn't deal with `including` rules or `denying` access

icehess committed:
fix rebase artifact

icehess committed:
normalize doc id for couchdb3

icehess committed:
add kapps_util:account_type/1 to detect account type

icehess committed:
fix spec

icehess committed:
revert and update view, use permission and account type

- Update most if the view.
- Add a new view to fetch scopes based on their account and user type

icehess committed:
fix available_scopes

icehess committed:
expand scopes during validation and store them in a process

icehess committed:
split scope rules

icehess committed:
revert ordering of funs to decrease diff

icehess committed:
import @jamesaimonetti changes

icehess committed:
cleanup a bit

icehess committed:
make apis and add missing kzd_scopes:id

icehess committed:
generic fetch rules and check http transport

icehess committed:
move precedence to scope_rule

icehess committed:
don't run scopes expand yet until specifically needed

icehess committed:
fix avialable_scopes again, use key since id is different

icehess committed:
use available_scopes as expander/resolver of superset scopes

icehess committed:
filter scope rules by application name

icehess committed:
store resolved/expanded scopes in process

icehess committed:
fix order in available scopes tests

The order is preseved now.

icehess committed:
flatten available_scopes when checking scopes

icehess committed:
rename CATCH_ALL macro

icehess committed:
add kz_auth_transport_websocket

icehess committed:
add catch all binding

icehess committed:
fix and update http transport, add tests

icehess committed:
add empty path

icehess committed:
basic eunit test for creating auth token

icehess committed:
split caller scopes from requested scopes

icehess committed:
fix include scope in claim

icehess committed:
eanup test

icehess committed:
export scope rule doc id fun

icehess committed:
make fmt

icehess committed:
rename function name

icehess committed:
fix

manushiperera commented:
Hello @icehess,
Can we ignore this related PR as well?

icehess commented:
yup

jamesaimonetti commented:
Ignoring for now, re-open on the other side if necessary
